src/datachain/toolkit/split.py

```diff
@@ -0,0 +1,66 @@
+from datachain import C, DataChain
+
+
+def train_test_split(dc: DataChain, weights: list[float]) -> list[DataChain]:
+    """
+    Splits a DataChain into multiple subsets based on the provided weights.
+
+    This function partitions the rows or items of a DataChain into disjoint subsets,
+    ensuring that the relative sizes of the subsets correspond to the given weights.
+    It is particularly useful for creating training, validation, and test datasets.
+
+    Args:
+        dc (DataChain): The DataChain instance to split.
+        weights (list[float]): A list of positive weights indicating the relative
+                               proportions of the splits.
+
+    Returns:
+        list[DataChain]:
+            A list of DataChain instances, one for each weight in the weights list.
+
+    Examples:
+        Train-test split:
+        ```python
+        from datachain import DataChain
+        from datachain.toolkit import train_test_split
+
+        # Load a DataChain from a storage source (e.g., S3 bucket)
+        dc = DataChain.from_storage("s3://bucket/dir/")
+
+        # Perform a 70/30 train-test split
+        train, test = train_test_split(dc, [0.7, 0.3])
+
+        # Save the resulting splits
+        train.save("dataset_train")
+        test.save("dataset_test")
+        ```
+
+        Train-test-validation split:
+        ```python
+        train, test, val = train_test_split(dc, [0.7, 0.2, 0.1])
+        train.save("dataset_train")
+        test.save("dataset_test")
+        val.save("dataset_val")
+        ```
+
+    Notes:
+        - The splits are random but deterministic, based on Dataset `sys__rand` field.
+        - The weights do not need to sum to 1; they will be normalized internally.
+          For example:
+            - `[0.7, 0.3]` corresponds to a 70/30 split;
+            - `[2, 1, 1]` corresponds to a 50/25/25 split;
+    """
+    if len(weights) < 2:
+        raise ValueError("Weights should have at least two elements")
+
+    weights_normalized = [round(weight * 1000 / sum(weights)) for weight in weights]
```

Comment:
Minor nitpick: consider making these values truly "normalized" in range 0-1, and moving the `* 1000` below, closer to the `%` operation.

Reply:
Also, I have been playing a bit with this normalization approach and sometimes I have found unexpected results:

```python
>>> sum(normalize([5, 5]))
1000
>>> sum(normalize([3, 3, 1]))
1001  # What!?
```

Not sure how important is that for this use case (maybe it's not), but looks like a potential subtle bug to me.

Reply:
I guess this is because of `round`. Have it removed.

Reply:
```
In [1]: weights = [3, 3, 1]

In [2]: [weight / sum(weights) for weight in weights]
Out[2]: [0.42857142857142855, 0.42857142857142855, 0.14285714285714285]

In [3]: sum([weight / sum(weights) for weight in weights])
Out[3]: 1.0
```